File: src/main/java/com/enderryno/nuclearcraft/Configuration/PluginConfiguration/PluginConfiguration.java
package com.enderryno.nuclearcraft.Configuration.PluginConfiguration;

import com.enderryno.nuclearcraft.Configuration.Enums.ConfigurationStorages;
import org.bukkit.configuration.file.FileConfiguration;
import org.bukkit.configuration.file.YamlConfiguration;
import org.bukkit.plugin.java.JavaPlugin;

import java.io.*;
import java.util.logging.Level;

public class PluginConfiguration {

    /* Strict constructor variables */
    private final JavaPlugin plugin;
    private final ConfigurationStorages configType;
    private final String filePath;
    private final String fileName;


    private File configurationFile = null;
    private File configurationFilePath = null;
    private FileConfiguration configuration = null;



    public PluginConfiguration(JavaPlugin plugin, String filePath, ConfigurationStorages configurationStorage) {
        this.plugin = plugin;
        this.configType = configurationStorage;

        this.filePath = filePath;
        this.fileName = configType.getStorageFileName() + ".yml";

        this.configurationFilePath = new File(plugin.getDataFolder(), filePath);
        this.configurationFile = new File(configurationFilePath, fileName);


        // Initialization of the default if it doesn't exist
        saveDefaultConfig();
    }

    public void reloadConfig() {
        if (this.configurationFilePath == null) {
            this.configurationFilePath = new File(plugin.getDataFolder(), filePath);
        }

        if (this.configurationFile == null) {
            this.configurationFile = new File(configurationFilePath, fileName);
        }

        this.configuration = YamlConfiguration.loadConfiguration(this.configurationFile);
        InputStream stream = plugin.getResource(fileName);
        if (stream != null) {
            YamlConfiguration YmlFile = YamlConfiguration.loadConfiguration(new InputStreamReader(stream));
            this.configuration.setDefaults(YmlFile);
        }


    }


    public void saveConfig() {
        try {
            this.configuration.save(this.configurationFile);
        } catch (IOException e) {
            this.plugin.getLogger().log(Level.SEVERE, "Could not save config to " + this.configurationFile, e);
        }
    }

    public void saveDefaultConfig() {
        if (!this.configurationFilePath.exists()) {
            if (!this.configurationFilePath.mkdirs()) {
                plugin.getLogger().log(Level.WARNING, "Could not create the directory!");
            }
        }

        if (!this.configurationFile.exists()) {
            plugin.saveResource(filePath + fileName, false);
        }


    }

    public FileConfiguration getConfig() {
        if (!configurationFilePath.exists()) {
            this.saveDefaultConfig();
        }

        if (configuration == null) {
            this.reloadConfig();
        }

        return this.configuration;

    }


    public static void save(PluginConfiguration config) {
        config.saveConfig();
        config.reloadConfig();
    }


}

File: src/main/java/com/mochibit/defcon/classes/PluginConfiguration.kt
/*
 *
 * DEFCON: Nuclear warfare plugin for minecraft servers.
 * Copyright (c) 2024 mochibit.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package com.mochibit.defcon.classes

import com.mochibit.defcon.enums.ConfigurationStorages
import org.bukkit.configuration.file.FileConfiguration
import org.bukkit.configuration.file.YamlConfiguration
import org.bukkit.plugin.java.JavaPlugin
import java.io.File
import java.io.IOException
import java.io.InputStreamReader
import java.util.logging.Level

class PluginConfiguration(private val plugin: JavaPlugin?, configurationStorage: ConfigurationStorages) {
    private val filePath: String = configurationStorage.storagePath
    private val fileName: String = configurationStorage.storageFileName + ".yml"
    private var configurationFile: File? = null
    private var configurationFilePath: File? = null
    private var configuration: FileConfiguration? = null

    init {
        configurationFilePath = File(plugin!!.dataFolder, filePath)
        configurationFile = File(configurationFilePath, fileName)

        // Initialization of the default if it doesn't exist
        saveDefaultConfig()
    }

    fun reloadConfig() {
        if (configurationFilePath == null) {
            configurationFilePath = File(plugin!!.dataFolder, filePath)
        }
        if (configurationFile == null) {
            configurationFile = File(configurationFilePath, fileName)
        }
        configuration = YamlConfiguration.loadConfiguration(configurationFile!!)
        val stream = plugin!!.getResource(fileName)
        if (stream != null) {
            val ymlFile = YamlConfiguration.loadConfiguration(InputStreamReader(stream))
            (configuration as YamlConfiguration).setDefaults(ymlFile)
        }
    }

    fun saveConfig() {
        try {
            configuration!!.save(configurationFile!!)
        } catch (e: IOException) {
            plugin!!.getLogger().log(Level.SEVERE, "Could not save config to $configurationFile", e)
        }
    }

    private fun saveDefaultConfig() {
        if (!configurationFilePath!!.exists()) {
            if (!configurationFilePath!!.mkdirs()) {
                plugin!!.getLogger().log(Level.WARNING, "Could not create the directory!")
                return
            }
        }
        if (!configurationFile!!.exists()) {
            plugin!!.saveResource(filePath + fileName, false)
        }
    }

    val config: FileConfiguration?
        get() {
            if (!configurationFilePath!!.exists()) {
                saveDefaultConfig()
            }
            if (configuration == null) {
                reloadConfig()
            }
            return configuration
        }

    companion object {
        fun save(config: PluginConfiguration) {
            config.saveConfig()
            config.reloadConfig()
        }
    }
}